torch/distributed/_composable/fsdp/_fsdp_param.py

```diff
@@ -0,0 +1,598 @@
+                Tuple[Placement, ...], tuple(global_placements)
+            )
+            self._global_size = self._tp_global_size
+            self._global_stride = self._tp_global_stride
+        else:
+            if _mesh_resources.get_parent_mesh(self.mesh_info.mesh) is not None:
+                raise NotImplementedError(
+                    "For 1D DP, using a parent mesh is not supported"
+                )
+            self._global_mesh = self.mesh_info.mesh
+            self._global_placements = (Shard(0),)
+            self._global_size = param.size()
+            self._global_stride = param.stride()
+
+    @torch.no_grad()
+    def _init_sharded_param(self, param: nn.Parameter, device: torch.device):
+        if param.device != device:
+            raise AssertionError(
+                "Expects parameter to already be moved to device "
+                f"{device} but got {param.device}"
+            )
+        param_data = param._local_tensor if isinstance(param, DTensor) else param
+        self._unsharded_size = param_data.size()
+        meta_param = torch.empty(
+            self._unsharded_size,
+            dtype=param.dtype,
+            layout=param.layout,
+            device="meta",
+            requires_grad=param.requires_grad,
+            # Assume contiguous memory layout
+        )
+        shard_rank = self.mesh_info.shard_mesh_rank
+        shard_world_size = self.mesh_info.shard_mesh_size
+        padded_param = pad_tensor_on_dim0(param_data, shard_world_size)
+        padded_chunks = torch.chunk(padded_param, shard_world_size, dim=0)
+        if self.offload_to_cpu:  # ignore `device`
+            self._sharded_param_data = param_data.new_empty(
+                padded_chunks[0].size(), device="cpu"
+            ).pin_memory()
+        else:
+            self._sharded_param_data = param_data.new_empty(
+                padded_chunks[0].size(), device=device
+            )
+        if device != torch.device("meta"):
+            self._sharded_param_data.copy_(padded_chunks[shard_rank])
+        chunks = chunk_with_empty(meta_param, shard_world_size, dim=0)
+        self._sharded_size = chunks[shard_rank].size()
+        self._padded_sharded_size = padded_chunks[0].size()
+        self._padded_unsharded_size = padded_param.size()
+        self._sharded_param_view = self._sharded_param_data[: self._sharded_size[0]]
+        self._sharded_param = nn.Parameter(
+            self.to_sharded_dtensor(self._sharded_param_view)
+        )
+        self._sharded_param.requires_grad_(param.requires_grad)
+        self._unsharded_param_data = param_data  # HACK: for `to_sharded()`
+        if not self.offload_to_cpu:
+            assert self._sharded_param.device == device
+        else:
+            assert self._sharded_param.device == torch.device("cpu")
+            self._cpu_sharded_grad = torch.empty_like(
+                self._sharded_param_view
+            ).pin_memory()  # no padding
```

Comment:
Even with `DTensor` improvements, I am not sure how we should handle the CPU sharded gradient cleanly.

The current CPU offloading implementation in this PR does not use pinned CPU memory (and hence no D2H copy overlap) if accumulating sharded gradients on CPU. If we want to support that, we actually need 2 pinned CPU sharded gradient tensors because if we do accumulate, the new sharded gradient must be copied to pinned memory plus the existing sharded gradient is in pinned memory (from the last D2H copy).

Pinning CPU memory on the fly can be expensive in terms of CPU overhead, so we prefer to pin at construction time.